### PR TITLE
[move-compiler-v2] Don't treat underscore parameters as variables in Move Lang V2+

### LINKED ISSUE #14949: [Bug][move-compiler-v2] Move allows `_` to be used on RHS, and treats `_` parameters as variables
# 🐛 Bug

While thinking about using `_` as a placeholder for some sort of curry operation, I found some strangeness in the way this special variable is handled in both V1 and V2.

- The value of a function parameter called `_` can be used.
- Having more than one function parameter called `_` leads to a "duplicate parameter" warning.
- `let` statements and assignments to `_` do not produce errors, but have no effect:
    - The code `let _ = 3; _` leads to an unbound/undeclared compile-time error in both compiler V1 and V2.
    - The program below shows assignments to `_` which are seen to have no effect on the value of `_` (which is obtained from the parameter, apparently not really the same thing as the LocalVar).  Essentially, assignment to `_` discards a value (as in Rust, actually).

For comparison, Rust says:
```
error: in expressions, `_` can only be used on the left-hand side of an assignment
 --> src/main.rs:4:5
  |
4 |     _
  |     ^ `_` not allowed here

```
and Rust ignores parameters called `_`, rather than generating errors due to multiple such parameters.

## Action

In Move Language V2 we should change this behavior to be more consistent:
- Disallow taking the value of `_`
- Ignore multiple parameters named `_`

This will free up `_` so we can use it for currying.

## Reproducing example

The below program demonstrates several of these problems:
```
//# publish
module 0xc0ffee::m {
    fun test2(): (u32, u64) {
        (1u32, 2u64)
    }
    
    public fun test(_: u64): u64 {
        let x = _ + 3;
        x + _
    }

    public fun test3(_: u64): u64 {
        let (_, _) = test2();
        let x = _ + 3;
        _ = _ + 1;
        x + _
    }

    public fun test4(_: u64): u64 {
        let (_, _x) = test2();
        _ = _ + 2;
        _
    }

    public fun test5(_: u64): u64 {
        let (_, _) = test2();
        _ = _ + 3;
        _
    }

    public fun test6(_: u64): u64 {
        let (_x, _) = test2();
         _
    }

    // public fun test7(_y: u64): u64 {
    //    let _ = _y;
    //    _  // undefined
    // }
}

//# run 0xc0ffee::m::test --args 4

//# run 0xc0ffee::m::test3 --args 5

//# run 0xc0ffee::m::test4 --args 5

//# run 0xc0ffee::m::test5 --args 5

//# run 0xc0ffee::m::test6 --args 5

//# run 0xc0ffee::m::test7 --args 5

```

Yields the following output, with no compiler errors (after `test7` was commented out):
```
processed 6 tasks

task 1 'run'. lines 37-37:
return values: 11

task 2 'run'. lines 39-39:
return values: 13

task 3 'run'. lines 41-41:
return values: 5

task 4 'run'. lines 43-43:
return values: 5

task 5 'run'. lines 45-45:
return values: 5

==> Compiler v2 delivered same results!
```

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

In Move Language 2.0+, treat a function parameter named "_" as a wildcard value-sink, not a variable.

This is needed if we want to use `_` in closure construction.

Fixes #14949.

## How Has This Been Tested?

Added more tests of underscore parameter and local vars.  Use the same
tests for V1 and compiler-v2 with Lang V1 and Lang V2.

- One or multiple params named _
- use of value of _
- Passing actual of wrong type to _ parameter
- Inline functions with same

Added duplicate transactional tests that comment out lines causing
compile-time errors (separately for V1 and V2) to demonstrate that
other transactions actually run correctly.

## Key Areas to Review

Did I miss any places where Parameters might be processed?

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [x] Move Compiler
- [ ] Other (specify)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-compiler-v2/transactional-tests/tests/operator_eval/underscore.operator-eval-lang-1.exp
Comment (wrwg) [inline third_party/move/move-compiler-v2/transactional-tests/tests/operator_eval/underscore.operator-eval-lang-1.exp:1]: Why this change require transactional tests? It's a poor compiler thing. The VM allows `_` as an identifier.

File: third_party/move/move-compiler-v2/transactional-tests/tests/operator_eval/underscore.operator-eval-lang-1.exp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+   ┌─ TEMPFILE:48:21
+   │
+48 │         fun10(4, 3, 2)
+   │                     ^
+
+
+
+task 1 'run'. lines 79-79:
+Error: Function execution failed with VMError: {
+    message: Linker Error: Module 0000000000000000000000000000000000000000000000000000000000c0ffee::m doesn't exist,
```
Comment (wrwg) [inline third_party/move/move-compiler-v2/transactional-tests/tests/operator_eval/underscore.operator-eval-lang-1.exp:45]: Kind of noisy unrelated followup errors.

File: third_party/move/move-compiler/tests/move_check/naming/underscore.exp
Comment (wrwg) [inline third_party/move/move-compiler/tests/move_check/naming/underscore.exp:1]: Not clear to me why we add tests to the v1 compiler.

**Review by wrwg — APPROVED** [review]
I'd prefer we keep our set of transactional tests focused on things were execution behavior is validated.

**Review by rahxephon89 — APPROVED** [review]
LGTM, thanks

File: third_party/move/move-compiler/tests/move_check/naming/underscore.exp
Comment (brmataptos) [reply third_party/move/move-compiler/tests/move_check/naming/underscore.exp:1]: It shows the behavior of the move code at runtime (Passing _ from the parameter, rather than the wildcard let pattern).

File: third_party/move/move-compiler-v2/transactional-tests/tests/operator_eval/underscore.operator-eval-lang-1.exp (hunk 1 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/transactional-tests/tests/operator_eval/underscore.operator-eval-lang-1.exp:45]: Its now a regression test.  No need to ever look at it again, unless it changes.
